source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+#pragma once
+
+#include <arpa/inet.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/network/address.h"
+
+#include "common/network/cidr_range.h"
+#include "common/network/utility.h"
+
+#include "absl/numeric/int128.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+/**
+ * Level Compressed Trie implemenation for IP(Ipv4 or IPv6) Prefixes.
+ */
+class LcTrie {
+public:
+  /**
+   * @param fill_factor supplies the fraction of completeness to use when considering
+   *        if a subtrie should be compressed.
+   * @param root_branching_factor supplies the branching factor at the root. This is
+   *        an optimization as suggested in the paper.
+   * @param tag_data supplies the ip tag and the prefixes for associated with that tag.
+   */
+  LcTrie(const std::vector<std::pair<std::string, std::vector<Address::CidrRange>>>& tag_data,
+         double fill_factor = 0.5, uint32_t root_branching_factor = 0);
+
+  /**
+   * Search the LC Trie for a prefix that has the input, ip, in range.
+   * @param  ip supplies the IP address searching for a prefix in string format.
+   * @return tag from the prefix that encompasses the input. An empty string is returned
+   *         if no prefix in the LC Trie exists.
+   */
+  std::string search(Network::Address::InstanceConstSharedPtr ip_address) const;
+
+private:
+  /**
+   * Extract n bits from input starting at position p.
+   * @param position supplies the position.
```

Comment:
It's called `position` in the comment but `p` in the declaration.

Reply:
Fixed.